ES|QL: Disable optimizations that rely on Expression.nullable()

luigidellaquila opened the pull request:
Fixes: https://github.com/elastic/elasticsearch/issues/104995

This PR disables all the planning optimizations that rely on `Expression.nullable()`. 
The semantics of this QL API have changed significantly in ESQL, so for now we'll just remove all the code that uses it. See also comments [here](https://github.com/elastic/elasticsearch/pull/105023#issuecomment-1955640677)

luigidellaquila committed:
Disable optimizations that rely on Expression.nullable()

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

luigidellaquila committed:
Update docs/changelog/105691.yaml

luigidellaquila committed:
Remove usage of nullable() from InferIsNotNull

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/esql/disable_nullable' into esql/disable_nullable

astefan reviewed (changes requested):
Since now ESQL is using its own rules for folding null and there is also the new PropagateNullable rule, you should look into adding many more tests to LogicalPlanOptimizerTests. Take a look at what's being tested in SQL (start from https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerTests.java#L342) and port over all the relevant tests and add new ones covering ESQL specific behavior.

luigidellaquila committed:
Add unit tests and simplify rule overrides

luigidellaquila commented:
Thanks @astefan, makes sense.
I added unit tests for all the rules I changed, and I realized that I could simplify some overrides.

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into esql/disable_nullable

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/FoldNull.java:
```diff
@@ -8,9 +8,8 @@
-import org.elasticsearch.xpack.ql.optimizer.OptimizerRules;
 
-public class FoldNull extends OptimizerRules.FoldNull {
+public class FoldNull extends LogicalPlanOptimizer.FoldNull {
```
I don't think this class is needed anymore.

luigidellaquila replied:
We still need it unfortunately, It's there because FoldNull.rule() is protected, and it's used also by `AbstractFuncitonTestCase`

luigidellaquila replied:
Unless we want to make it public in ESQL

astefan reviewed (commented):
The tests you added are a good start, but they don't seem to cover what the OptimizerTests was covering in SQL. One example is `testGenericNullableExpression` which looks at more types of expressions than this PR is. There is also `testNullFoldingDoesNotApplyOnConditionals` that deals with conditional functions (ESQL has these as well).

The two tests above are just two examples, I didn't make an exhaustive search for what's missing and what not. Please, take your time to address all the relevant tests.

luigidellaquila committed:
More tests

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into esql/disable_nullable

luigidellaquila committed:
Merge branch 'main' into esql/disable_nullable

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -3312,4 +3366,279 @@ public static RLike rlike(Expression left, String exp) {
+        And top = new And(EMPTY, and, lessThanOf(fa, ONE));
+
+        Expression optimized = new LogicalPlanOptimizer.PropagateNullable().rule(top);
+        Expression expected = new And(EMPTY, and, nullOf(BOOLEAN));
```
Actually, running this through a test (`row a = 1, b = 0, c = 0 | eval a IS NULL AND b < 1 AND c < 1 AND a <= 1`) the plan is optimized as `[ISNULL(1[INTEGER]) AND null[BOOLEAN] AND null[BOOLEAN] AND null[BOOLEAN]`. This is not right. For whatever reason (I stopped looking into it at this point), `1` (as Literal) is added to the list of `nullExpressions` in `OptimizerRules.PropagateNullable.rule` and then, with your change in `LogicalPlanOptimizer.PropagateNullable.nullify` the `1` (Literal) is replaced as the "null" value in all other expressions where 1 is encountered.

For a different test where the coincidence of `1` disappears (`a = 1` and `a is null` and `b < 1`.... that is the same `1`) the rule yields something else. So the test is `row a = 5, b = 0, c = 0 | eval a IS NULL AND b < 1 AND c < 1 AND a <= 1` and the rule PropagateNullable changes the plan to `[ISNULL(5[INTEGER]) AND 0[INTEGER] < 1[INTEGER] AND 0[INTEGER] < 1[INTEGER] AND null[BOOLEAN]`.

I see two problems, that all might have to do with a single source:
- `1` gets replaced in the expressions as `null`
- `a <= 1` becomes `null`

luigidellaquila replied:
That's interesting, I'll look deeper into this case, maybe I broke something in the original logic

luigidellaquila replied:
The problem seems to be due to an interference between two rules ([PropagateEvalFoldables](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java#L479) and [PropagateNullable](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1705)), my change only exposed it. 
In addition, the different semantics of nullability in ESQL makes this whole management much more complex.

What happens is the following:
1. `PropagateEvalFoldables` replaces all the occurrences of `a` and `b` with `1`, so `row a = 1, b = 1, c = 1 | eval a IS NULL AND b < 1 AND c < 1 AND a <= 1` becomes `row a = 1, b = 1, c = 1 | eval 1 IS NULL AND 1 < 1 AND 1 < 1 AND 1 <= 1`. This is not evident in the planning logs, because when we print the single plans we only see the original source, not the actual values in the expression nodes (see screenshot below)
2. `PropagateNullable` tries to optimize `a IS NULL ... AND a <= 1` as `a IS NULL ... AND NULL <= 1` (and not `a IS NULL AND FALSE`, the comment [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1703) is wrong. The replacer used in the rule is [this::nullify](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1741), that produces `NULL` for the attribute, not `FALSE` for the whole expression). 
3. This happens by [finding the field](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1722) that is supposed to be null (`a` in this case), looking for it in other expressions, and [replacing it](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1741) with `NULL`. The problem is that there is no `a` at this time, because it was replaced with `1` (see point 1). Even worse, also other expressions were replaced with `1` (in this case, `b < 1` is now `NULL < 1`), so they will also be affected by this replacement, that is completely unexpected.

See this screenshot: when trying to infer the null expressions, the actual field value `isn.field()`, that was expected to return `a`, returns `1` instead

<img width="683" alt="Screenshot 2024-03-05 alle 14 08 16" src="https://github.com/elastic/elasticsearch/assets/459289/69cca2b0-080d-4d05-aa48-90b168db58a2">

---

So why did it work before?

In the original implementation it worked only because expressions like `constant < constant` were considered non-nullable, so `nullify()` returned the original expression. In practice, the rule did nothing in this case.

IMHO this is another problem: [relying on nullability to replace values with NULL](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1782) does not seem to be so safe: if we know that `a == 1` and we try to optimize `a < 0`, we cannot just replace `a` with null, because calculating  `1 < 0` is different from calculating `null < 0`, due to null virality.

--- 

How to fix it?

The problem seems to be limited to foldable literals, so we could explicitly exclude this case. Something like

```java
        protected Expression nullify(Expression exp, Expression nullExp) {
            return exp.foldable() ? exp : Literal.of(exp, null);
        }
```
An alternative is to completely drop this rule, that would be the safest path, but also a shame because it looks correct (though a bit intricate)  for non-foldable attributes, so we would give up on a valid optimization.

Or we could try to completely rewrite it, but it doesn't seem trivial.

Suggestions are welcome

astefan replied:
My alternative from a month ago was to split FoldNull: https://github.com/elastic/elasticsearch/compare/main...astefan:elasticsearch:fold_null_split?expand=1.

luigidellaquila replied:
Turns out that the fix above on `nullify()` method is not enough: if a constant value becomes a target for "nullification", it will be transformed in other unexpected places, eg. in the right-hand side of a completely unrelated expression.

I went for a different solution: exclude constant expressions from `PropagateNullable` logic since the beginning. 
In terms of performance it should not make any difference, foldable expressions will eventually be folded anyway, there is no need to nullify them in advance.
The case you mentioned above should be fine now, I added a specific test for that

costin replied:
> The replacer used in the rule is [this::nullify](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java?rgh-link-date=2024-03-05T15%3A05%3A13Z#L1741), that produces NULL for the attribute, not FALSE for the whole expression).

The original method returns null IF the expression is nullable which is true for functions only if at least one child is null. However it is also true for attributes / FieldAttributes leading to its replacement with null.
In might be that in the SQL planner this didn't get a chance kick in because of the other rules - one fix would be to change the nullify method.

astefan reviewed (commented):
Better tests this time. Thanks.
But, there are some fine details in these tests that reveal something that doesn't work right. Please, see my comments in LogicalPlanOptimizerTests.

luigidellaquila commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -3312,4 +3366,279 @@ public static RLike rlike(Expression left, String exp) {
+        FieldAttribute fa = getFieldAttribute("a");
+
+        And andOne = new And(EMPTY, new IsNull(EMPTY, fa), new IsNotNull(EMPTY, getFieldAttribute("a")));
+        And andTwo = new And(EMPTY, new IsNull(EMPTY, getFieldAttribute("a")), new IsNotNull(EMPTY, getFieldAttribute("a")));
```
Good catch, let me fix it

luigidellaquila commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -3312,4 +3366,279 @@ public static RLike rlike(Expression left, String exp) {
+        IsNull isNull = new IsNull(EMPTY, fa);
+
+        And and = new And(EMPTY, isNull, greaterThanOf(fa, ONE));
+        assertEquals(new And(EMPTY, isNull, nullOf(BOOLEAN)), new LogicalPlanOptimizer.PropagateNullable().rule(and));
```
oh yes, the comment is incorrect

luigidellaquila committed:
Tests cleanup

luigidellaquila committed:
Ignore foldable expressions on null propagation

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1587,4 +1589,72 @@ private static LogicalPlan normalize(Aggregate aggregate, AttributeMap<Expressio
+            return and;
+        }
+
+        protected Expression nullify(Expression exp, Expression nullExp) {
+            return Literal.of(exp, null);
+        }
```
This doesn't look right - to always return null regardless of the nullExp.

luigidellaquila replied:
I see what you mean, it seems a stretch, but:
- the parent method returns null regardless of the nullExp as well
- I cannot speculate on expression nullability, because we know in ESQL the nullability logic is inconsistent

SQL only uses `nullExp` to do one additional optimization on `coalesce()` (remove the expressions that we know are null); I'll add this optimization, but it doesn't change the correctness of the rule.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1587,4 +1589,72 @@ private static LogicalPlan normalize(Aggregate aggregate, AttributeMap<Expressio
+
+            // first find isNull/isNotNull
+            for (Expression ex : splits) {
+                // Avoid considering constants. It would lead to collisions with foldable propagation
+                // Folding will take care of them eventually anyway
+                if (ex instanceof IsNull isn && isn.field().foldable() == false) {
+                    nullExpressions.add(isn.field());
+                } else if (ex instanceof IsNotNull isnn && isnn.field().foldable() == false) {
+                    notNullExpressions.add(isnn.field());
+                }
```
Why is there a collision over foldables?

luigidellaquila replied:
I explained it here https://github.com/elastic/elasticsearch/pull/105691#discussion_r1512986433

The problem is that `PropagateEvalFoldables` (that only exists in ESQL, that is probably one of the reasons why we don't have this problem in SQL) replaces fields with their value when they are foldable, so:
- if `a = 1`, an expression like `a is null` could become `1 is null`
- for `a is null`, that now is `1 is null`, `isn.field()` returns `1`, that gets added to nullExpressions
- nullExpressions are then replaced with NULL in other expressions in the same AND block, so also completely unrelated expressions like `b > 1` could be impacted, that is exactly the problem @astefan reported

In SQL the problem doesn't happen also because `nullify()` can rely on `nullable() = false` to filter out these cases. Paradoxically, if we cannot rely on the nullability logic in ESQL, we don't even know that `1` is nullable.

luigidellaquila replied:
> Paradoxically, if we cannot rely on the nullability logic in ESQL, we don't even know that 1 is not nullable.

in this specific case I could fold, but we have other rules that do that, this one is about nullability

costin reviewed (changes requested):
Great find by @astefan on null rules. Left a few comments to help minimize the amount of code in this PR.

costin reviewed (changes requested):
Great find by @astefan on null rules. Left a few comments to help minimize the amount of code in this PR.

luigidellaquila commented:
> The original method returns null IF the expression is nullable which is true for functions only if at least one child is null. However it is also true for attributes / FieldAttributes leading to its replacement with null.

@costin the problem is that we cannot rely on nullability logic. In ESQL an expression can be null also in other situations, eg. mv or number overflows.

> In might be that in the SQL planner this didn't get a chance kick in because of the other rules - one fix would be to change the nullify method.

The problem doesn't kick in in SQL for two reasons:
- in ESQL we have `PropagateEvalFoldables` that injects constants when SQL had attributes
- SQL can rely on nullability logic, ESQL cannot

I tried to change `nullify()`, but it didn't work: at that stage it's too late, I tried to summarize it in the [comments above](https://github.com/elastic/elasticsearch/pull/105691#discussion_r1515823851)

luigidellaquila committed:
Implement review suggestions

luigidellaquila committed:
Merge branch 'main' into esql/disable_nullable

luigidellaquila committed:
Fix checkstyle

costin committed:
Swap folding rule order

Perform propagation of foldables before constant folding - this helps
 evaluate early plans that define constants. In this case triggers
 early evaluation of isnull/isnotnull inside PropagateNullable.

luigidellaquila committed:
Fix bwc tests

luigidellaquila commented:
@elasticmachine run elasticsearch-ci/rest-compatibility

luigidellaquila commented:
`elasticsearch-ci/rest-compatibility` finished and was green, but for some reason  GitHub after a few hours still didn't see it...

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1848,10 +1856,10 @@ protected Set<Expression> resolveExpressionAsRootAttributes(Expression exp, Attr
         }
 
-        private boolean doResolve(Expression exp, AttributeMap<Expression> aliases, Set<Expression> resolvedExpressions) {
+        protected boolean doResolve(Expression exp, AttributeMap<Expression> aliases, Set<Expression> resolvedExpressions) {
```
The method can be kept private.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1752,7 +1760,7 @@ protected Expression rule(And and) {
          * or removed if null is returned.
          */
-        private static boolean replace(
+        protected static boolean replace(
```
Method can be kept private.

costin reviewed (approved):
LGTM

astefan reviewed (approved):
LGTM

luigidellaquila committed:
Implement review suggestions

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.13](https://github.com/elastic/elasticsearch/pull/106132)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/106132">](https://github.com/elastic/elasticsearch/pull/106132) |
